src/main/kotlin/org/jitsi/videobridge/transport/ice/IceTransport.kt

```diff
@@ -0,0 +1,444 @@
+
+        val remoteCandidates = transportPacketExtension.getChildExtensionsOfType(CandidatePacketExtension::class.java)
+        if (iceAgentStateIsRunning && remoteCandidates.isEmpty()) {
+            logger.cdebug { "Ignoring transport extensions with no candidates, " +
+                    "the Agent is already running." }
+            return
+        }
+
+        val remoteCandidateCount = addRemoteCandidates(remoteCandidates, iceAgentStateIsRunning)
+        if (iceAgentStateIsRunning) {
+            when (remoteCandidateCount) {
+                0 -> {
+                    // XXX Effectively, the check above but realizing that all
+                    // candidates were ignored:
+                    // iceAgentStateIsRunning && candidates.isEmpty().
+                }
+                else -> iceComponent.updateRemoteCandidates()
+            }
+        } else if (remoteCandidateCount != 0) {
+            // Once again, because the ICE Agent does not support adding
+            // candidates after the connectivity establishment has been started
+            // and because multiple transport-info JingleIQs may be used to send
+            // the whole set of transport candidates from the remote peer to the
+            // local peer, do not really start the connectivity establishment
+            // until we have at least one remote candidate per ICE Component.
+            if (iceComponent.remoteCandidateCount > 0) {
+                logger.info("Starting the agent with remote candidates.")
+                iceAgent.startConnectivityEstablishment()
+            }
+        } else if (iceStream.remoteUfragAndPasswordKnown()) {
+            // We don't have any remote candidates, but we already know the
+            // remote ufrag and password, so we can start ICE.
+            logger.info("Starting the Agent without remote candidates.")
+            iceAgent.startConnectivityEstablishment()
+        } else {
+            logger.cdebug { "Not starting ICE, no ufrag and pwd yet. ${transportPacketExtension.toXML()}" }
+        }
+    }
+
+    fun startReadingData() {
+        logger.cdebug { "Starting to read incoming data" }
+        val socket = iceComponent.socket
+        val receiveBuf = ByteBufferPool.getBuffer(1500)
+        val packet = DatagramPacket(receiveBuf, 0, receiveBuf.size)
+        var receivedTime: Instant
+
+        while (running.get()) {
+            try {
+                socket.receive(packet)
+                receivedTime = clock.instant()
+            } catch (e: SocketClosedException) {
+                logger.info("Socket closed, stopping reader")
+                break
+            } catch (e: IOException) {
+                logger.warn("Stopping reader", e)
+                break
+            }
+            packetStats.numPacketsReceived++
+            incomingDataHandler?.dataReceived(receiveBuf, 0, packet.length, receivedTime) ?: run {

```

Comment:
Is it possible that `receive` set a non-zero offset?

Reply:
I dunno, but agree it would be best to use `packet.offset` (I think I forgot it had one).